pkg/operator/controller/instance_controller.go

```diff
@@ -0,0 +1,742 @@
+											FieldPath: "metadata.namespace",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	return r.createOrUpdateResource(ctx, cpDeployment)
+}
+
+// setDataplane sets up the dataplane deployment.
+func (r *InstanceReconciler) setDataplane(ctx context.Context) error {
+	DataplaneImage := DataPlaneName
+	if r.instance.Spec.DataPlane.Type == api.DataplaneTypeGo {
+		DataplaneImage = GoDataPlaneName
+	}
+
+	dpDeployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      DataPlaneName,
+			Namespace: r.clNamespace,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: int32Ptr(int32(r.instance.Spec.DataPlane.Replicas)),
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{"app": DataPlaneName},
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{"app": DataPlaneName},
+				},
+				Spec: corev1.PodSpec{
+					Volumes: []corev1.Volume{
+						{
+							Name: "ca",
+							VolumeSource: corev1.VolumeSource{
+								Secret: &corev1.SecretVolumeSource{
+									SecretName: "cl-fabric",
+								},
+							},
+						},
+						{
+							Name: "tls",
+							VolumeSource: corev1.VolumeSource{
+								Secret: &corev1.SecretVolumeSource{
+									SecretName: DataPlaneName,
+								},
+							},
+						},
+					},
+					Containers: []corev1.Container{
+						{
+							Name:  "dataplane",
+							Image: r.instance.Spec.ContainerRegistry + DataplaneImage + ":" + r.instance.Spec.ImageTag,
+							Args: []string{
+								"--log-level", r.instance.Spec.LogLevel,
+								"--controlplane-host", ControlPlaneName,
+							},
+							ImagePullPolicy: corev1.PullIfNotPresent,
+							Ports: []corev1.ContainerPort{
+								{
+									ContainerPort: dpapi.ListenPort,
+								},
+							},
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									Name:      "ca",
+									MountPath: dpapp.CAFile,
+									SubPath:   "ca",
+									ReadOnly:  true,
+								},
+								{
+									Name:      "tls",
+									MountPath: dpapp.CertificateFile,
+									SubPath:   "cert",
+									ReadOnly:  true,
+								},
+								{
+									Name:      "tls",
+									MountPath: dpapp.KeyFile,
+									SubPath:   "key",
+									ReadOnly:  true,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	return r.createOrUpdateResource(ctx, dpDeployment)
+}
+
+// setService sets up a k8s service.
+func (r *InstanceReconciler) setService(ctx context.Context, name string, port uint16) error {
+	service := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: r.clNamespace},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Protocol: corev1.ProtocolTCP,
+					Port:     int32(port),
+				},
+			},
+			Type:     corev1.ServiceTypeClusterIP,
+			Selector: map[string]string{"app": name},
+		},
+	}
+
+	return r.createResource(ctx, service)
```

Comment:
any reason not to use `createOrUpdateResource` here as done in the other places?